core/blockchain.go

```diff
@@ -899,8 +907,14 @@ func (bc *BlockChain) WriteBlockAndState(block *types.Block, receipts []*types.R
 			bc.procTime = 0
 		}
 		// Garbage collect anything below our required write retention
-		bc.triedb.Dereference(header.Root, common.Hash{})
-
+		for !bc.triegc.Empty() {
+			root, number := bc.triegc.Pop()
+			if uint64(-number) > chosen {
```

Comment:
Since we are going to lose precision here after 16M blocks (which seems a bit ugly but acceptable for the time being), I'd at least make this comparison err on the safe side in order to never delete anything newer than `chosen`:
```
if -number >= float32(chosen+1) {
```